File: go/606.go
package main

import (
	"strconv"
	"strings"
)

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func tree2str(root *TreeNode) string {
	var ret []string = make([]string, 0)
	var helper func(t *TreeNode)
	helper = func(t *TreeNode) {
		if t == nil {
			return
		}
		ret = append(ret, strconv.Itoa(t.Val))
		if t.Left != t.Right {
			ret = append(ret, "(")
			helper(t.Left)
			ret = append(ret, ")")
			if t.Right != nil {
				ret = append(ret, "(")
				helper(t.Right)
				ret = append(ret, ")")
			}
		}
	}
	helper(root)
	return strings.Join(ret, "")
}

File: cpp/606.cpp
#include <iostream>
#include <string>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
class Solution {
public:
  void helper(TreeNode *t, string &s) {
    if (!t)
      return;
    s += to_string(t->val);
    if (t->left || t->right) {
      s += "(";
      helper(t->left, s);
      s += ")";
      if (t->right) {
        s += "(";
        helper(t->right, s);
        s += ")";
      }
    }
  }
  string tree2str(TreeNode *root) {
    string ret = "";
    helper(root, ret);
    return ret;
  }
};
